core/src/main/java/com/google/instrumentation/stats/MutableDistribution.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2017, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.instrumentation.stats;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Mutable version of {@link Distribution}.
+ */
+public final class MutableDistribution {
+  private long count = 0; // The number of values in the population.
+  private double sum = 0.0; // The sum of the values in the population.
+  private Range range = Range.create(); // Range of values in the population.
+
+  @Nullable
+  private final BucketBoundaries bucketBoundaries; // Histogram boundaries; null means no histogram
+
+  @Nullable
+  private long[] bucketCounts; // Counts for each histogram bucket
+
+  /**
+   * Constructs a new, empty {@link MutableDistribution}.
+   *
+   * @return a new, empty {@code MutableDistribution}.
+   */
+  public static final MutableDistribution create() {
+    return new MutableDistribution(null);
+  }
+
+  /**
+   * Constructs a new {@link MutableDistribution} with specified {@code bucketBoundaryList}.
+   *
+   * @param bucketBoundaries the boundaries for the buckets in the underlying {@code
+   *     MutableDistribution}.
+   * @return a new, empty {@code MutableDistribution} with the specified boundaries.
+   */
+  public static final MutableDistribution create(BucketBoundaries bucketBoundaries) {
+    checkNotNull(bucketBoundaries, "bucketBoundaries Object should not be null.");
+    return new MutableDistribution(bucketBoundaries);
+  }
+
+  // Returns true if the distribution has histogram buckets.
+  private boolean hasBuckets() {
+    return (bucketBoundaries != null);
+  }
+
+  // Construct a new MutableDistribution with optional bucket boundaries.
+  private MutableDistribution(@Nullable BucketBoundaries bucketBoundaries) {
+    this.bucketBoundaries = bucketBoundaries;
+    if (hasBuckets()) {
```

Comment:
I feel that is better in the constructor to not call hasBuckets() because it may change to be bucketCounts != null and the code will break.

bucketCounts = bucketBoundaries != null ? new long[bucketBoundaries.getBoundaries().size() + 1] : null;